File: clojure/rna-transcription/src/rna_transcription.clj
(ns rna-transcription)

(defn dna->rna
  "Translates a single character representing a DNA nucleotide into its
   corresponding RNA nucleotide.

   Throws an assertion error if the argument is not a valid DNA nucleotide."
  [nucleotide]
  {:post [(not (nil? %))]}
  (let [lookup (zipmap "GCTA" "CGAU")]
    (lookup nucleotide)))

(defn to-rna [strand]
  (apply str (map dna->rna strand)))

File: ruby/rna-transcription/rna_transcription.rb
module BookKeeping
  VERSION = 4
end

module Complement
  DNA_TO_RNA = {
    'G' => 'C',
    'C' => 'G',
    'T' => 'A',
    'A' => 'U'
  }

  def Complement.of_dna(strand)
    strand.each_char.with_object('') do |dna, rna|
      nucleotide = DNA_TO_RNA[dna]
      return '' unless nucleotide
      rna << nucleotide
    end
  end
end
